lib/vector-config/src/lib.rs

```diff
@@ -0,0 +1,343 @@
+// High-level list of TODOS.
+//
+// TODO: `serde` supports defining a default at the struct level to fill in fields when no value is present during
+// serialization, but it also supports defaults on a per-field basis, which override any defaults that would be applied
+// by virtue of the struct-level default.
+//
+// Thus, we should mark a struct optional if it has a struct-level default _or_ if all fields are optional: either
+// literal `Option<T>` fields or if they all have defaults.
+//
+// This could clean up some of the required properties where we have a field-level/struct-level default that we can
+// check by looking at the metadata for the type implementing `T`, maybe even such that the default impl of
+// `Configurable::is_optional` could just use that.
+//
+// TODO: What happens if we try to stick in a field that has a struct with a lifetime attached to it? How does the name
+// of that get generated in terms of what ends up in the schema? Do we even have fields with lifetime bounds in any of
+// our configuration types in `vector`? :thinking:
```

Comment:
I can't think of any config types that have lifetime bounds. I don't really know why they'd ever need them.

Reply:
Yeah, fair point.